### PR TITLE
using lzma to compress/decompress state_parts

### LINKED ISSUE #1048: Compress state at StateSync
Use ZIP or similar to compress the state when sending.
 Also optimize the sending to send state in parts / collect them on the other side.

### PR DESCRIPTION
**Pull request by Kouprin**
Fixes #1048.

### CODE REVIEW

File: chain/chain/src/chain.rs [hunk 1]
```diff
@@ -1390,19 +1391,36 @@ impl Chain {
...
+                // Saving the part data.
+                let mut store_update = self.store.owned_store().store_update();
+                let key = StatePartKey(sync_hash, shard_id, part_id).try_to_vec()?;
+                store_update.set_ser(COL_STATE_PARTS, &key, &decompressed_data)?;
+                store_update.commit()?;
+            }
+            _ => {
+                byzantine_assert!(false);
+                return Err(ErrorKind::Other(
+                    "set_state_part failed: lzma::decompress failed".into(),
```
Comment (bowenwang1996) [inline chain/chain/src/chain.rs]: Why not show the decompression error here? Might help with debugging

**Comment by mikhailOK** [comment]
is it possible to create a decompression bomb for lzma?

File: chain/chain/src/chain.rs (hunk 1 above)
Comment (Kouprin) [reply chain/chain/src/chain.rs]: It should not fail at all.

Comment (bowenwang1996) [reply chain/chain/src/chain.rs]: So every byte sequence can be decoded?

**Comment by Kouprin** [comment]
https://en.wikipedia.org/wiki/Zip_bomb

**Comment by Kouprin** [comment]
https://bomb.codes/bombs
Should we write our own compressor then?

File: chain/chain/src/chain.rs (hunk 1 above)
Comment (Kouprin) [reply chain/chain/src/chain.rs]: Not failing means that failing caused by clearly malicious data.

**Comment by mikhailOK** [comment]
We need a library that allows setting a resource limit for decompression + somehow know an upper bound for the size of one part

**Comment by Kouprin** [comment]
> We need a library that allows setting a resource limit for decompression + somehow know an upper bound for the size of one part

Rust-lzma allows to limit memory usage, as usually others do. I'm trying to find a library which allows to set limited output buffer.

We have already limited upper bound of one part with 1 Mb unpacked data.

**Comment by mikhailOK** [comment]
Can we guarantee 1 MB limit? The logic for splitting is approximate

**Comment by Kouprin** [comment]
> Can we guarantee 1 MB limit? The logic for splitting is approximate

Trie relies on `memory_usage` field while splitting. Are we giving enough room to guarantee that the size of the part is no more than X Mb?

**Comment by mikhailOK** [comment]
we might be able to guarantee that for some X after the coming storage changes + setting a limit on value sizes

**Comment by Kouprin** [comment]
> we might be able to guarantee that for some X after the coming storage changes + setting a limit on value sizes

Yep, it's important. Should we create a task for it?

**Comment by mikhailOK** [comment]
created #1786 , for now we should set max size to something big  - probably not infinite because of bombs, but potentially no smaller than the whole state size

**Comment by Kouprin** [comment]
What number you suggest to set for now?

**Comment by mikhailOK** [comment]
`state_root.memory_usage * 10 + 1_000_000`

**Comment by Kouprin** [comment]
> `state_root.memory_usage * 10 + 1_000_000`

I expect it should be a constant.

Anyway, we can wait till #1792 will be accepted.

**Comment by mikhailOK** [comment]
After #1792 we still need limit on keys and values

**Comment by ilblackdragon** [comment]
Is this PR still alive?

**Comment by Kouprin** [comment]
doesn't seem actual for now

**Pull request closed without merging**